File: TeXmacs/examples/plugins/pycomplete/src/complete.py
#!/usr/bin/env python
###############################################################################
##
## MODULE      : complete.py
## DESCRIPTION : Complete implemented in Python
## COPYRIGHT   : (C) 2019  Darcy Shen
##
## This software falls under the GNU general public license version 3 or later.
## It comes WITHOUT ANY WARRANTY WHATSOEVER. For details, see the file LICENSE
## in the root directory or <http://www.gnu.org/licenses/gpl-3.0.html>.

import os
import sys
from os.path import exists
tmpy_home_path = os.environ.get("TEXMACS_HOME_PATH") + "/plugins/tmpy"
if (exists (tmpy_home_path)):
    sys.path.append(os.environ.get("TEXMACS_HOME_PATH") + "/plugins/")
else:
    sys.path.append(os.environ.get("TEXMACS_PATH") + "/plugins/")

from tmpy.protocol        import *
from tmpy.compat          import *

flush_verbatim ("We know how to complete 'h'")

while True:
    line = tm_input()
    if not line:
        pass
    if line[0] == DATA_COMMAND:
        flush_scheme ("""(tuple "h" "ello" "i there" "ola" "opsakee")""")
    else:
        flush_verbatim ("You typed " + line)

File: src/TeXmacs/examples/plugins/complete/src/complete.cpp

/******************************************************************************
* MODULE     : complete.cpp
* DESCRIPTION: Shows how to program tab-completion
* COPYRIGHT  : (C) 2003  Joris van der Hoeven
*******************************************************************************
* This software falls under the GNU general public license and comes WITHOUT
* ANY WARRANTY WHATSOEVER. See the file $TEXMACS_PATH/LICENSE for more details.
* If you don't have this file, write to the Free Software Foundation, Inc.,
* 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
******************************************************************************/

#include <stdio.h>
#include <iostream.h>

#define DATA_BEGIN   ((char) 2)
#define DATA_END     ((char) 5)
#define DATA_COMMAND ((char) 16)
#define DATA_ESCAPE  ((char) 27)

void
format_plugin () {
  // The configuration of a plugin can be completed at startup time.
  // This is for instance interesting if you add tab-completion a posteriori.
  cout << DATA_BEGIN << "command:";
  cout << "(plugin-configure complete (:tab-completion #t))";
  cout << DATA_END;
}

int
main () {
  cout << DATA_BEGIN << "verbatim:";
  format_plugin ();
  cout << "We know how to complete 'h'";
  cout << DATA_END;
  fflush (stdout);

  while (true) {
    char buffer[100];
    cin.getline (buffer, 100, '\n');
    if (buffer[0] != DATA_COMMAND) {
      cout << DATA_BEGIN << "verbatim:";
      cout << "You typed " << buffer;
      cout << DATA_END;
    }
    else {
      cout << DATA_BEGIN << "scheme:";
      cout << "(tuple \"h\" \"ello\" \"i there\" \"ola\" \"opsakee\")";
      cout << DATA_END;
    }
    fflush (stdout);
  }
  return 0;
}
